src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.Value.cs

```diff
@@ -0,0 +1,264 @@
+                    return true;
+                }
+
+                osPlatformName = null;
+                return false;
+            }
+
+            private static bool TryDecodeOSVersion(
+                ImmutableArray<IArgumentOperation> arguments,
+                ValueContentAnalysisResult? valueContentAnalysisResult,
+                [NotNullWhen(returnValue: true)] out Version? osVersion,
+                int skip = 0)
+            {
+
+                using var versionBuilder = ArrayBuilder<int>.GetInstance(4, fillWithValue: 0);
+                var index = 0;
+
+                foreach (var argument in arguments.Skip(skip))
+                {
+                    if (!TryDecodeOSVersionPart(argument, valueContentAnalysisResult, out var osVersionPart))
+                    {
+                        osVersion = null;
+                        return false;
+                    }
+
+                    versionBuilder[index++] = osVersionPart;
+                }
+
+                osVersion = CreateVersion(versionBuilder);
+
+                return true;
+
+                static bool TryDecodeOSVersionPart(IArgumentOperation argument, ValueContentAnalysisResult? valueContentAnalysisResult, out int osVersionPart)
+                {
+                    if (argument.Value.ConstantValue.HasValue &&
+                        argument.Value.ConstantValue.Value is int versionPart)
+                    {
+                        osVersionPart = versionPart;
+                        return true;
+                    }
+
+                    if (valueContentAnalysisResult != null)
+                    {
+                        var valueContentValue = valueContentAnalysisResult[argument.Value];
+                        if (valueContentValue.IsLiteralState &&
+                            valueContentValue.LiteralValues.Count == 1 &&
+                            valueContentValue.LiteralValues.Single() is int part)
+                        {
+                            osVersionPart = part;
+                            return true;
+                        }
+                    }
+
+                    osVersionPart = default;
+                    return false;
+                }
+
+                static Version CreateVersion(ArrayBuilder<int> versionBuilder)
+                {
+                    if (versionBuilder[3] == 0)
+                    {
+                        if (versionBuilder[2] == 0)
+                        {
+                            return new Version(versionBuilder[0], versionBuilder[1]);
+                        }
+                        else
+                        {
+                            return new Version(versionBuilder[0], versionBuilder[1], versionBuilder[2]);
+                        }
+                    }
+                    else
+                    {
+                        return new Version(versionBuilder[0], versionBuilder[1], versionBuilder[2], versionBuilder[3]);
+                    }
```

Comment:
Be careful. `Version` compares funny. You may want to pad with zeros instead.

Reply:
+1 on zero-padding; I think that was mentioned in one of the API reviews that we would need to.

Reply:
The difference between our version check and the `Version` type was caused from `Revision, Build` fields defaulted to -1, with this conditions instead of creating Version(major, manor, 0, 0) creating it Version(major, manor) so that defaults applied, everywhere else we use default for creating versions, so this allows us to use `>, <, >=, <=` operators of `Version` type directly